### PR TITLE
feat: [PLG-2233] Add release version support for STO

### PR DESCRIPTION
**Pull request by neelam-harness**
## Describe your changes
1. Added support for fetching the STO version number in Release Notes
2. Refactoring the code by adding switch-case instead of if-else cases and error handling with try-catch blocks

Tested locally:
1. Ran the sto-core server locally by following https://github.com/harness/sto-core#readme 
2. successfully GET request locally to get the version in the response
image
3. Updated the moduleVersionInfo collection's document for STO as follows:
image

Note:
Gotchas for running the postgres in local for sto-core
1. `docker pull postgres:14.4`
2. `docker run -d --name postgresSto -p 5432:5432 -e POSTGRES_PASSWORD=sto_local postgres:14.4`
3. `docker exec postgresSto sh -c 'echo "DO \$\$ BEGIN IF NOT EXISTS(SELECT schema_name FROM information_schema.schemata WHERE schema_name = '"'"'sto'"'"') THEN EXECUTE '"'"'CREATE SCHEMA sto'"'"'; END IF; END \$\$ ;" | psql postgres://postgres:sto_local@localhost:5432/'`

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52046)

### CODE REVIEW

File: 120-ng-manager/config.yml [hunk 1]
```diff
@@ -142,6 +142,11 @@ cvngClientConfig:
   connectTimeOutSeconds: 15
   readTimeOutSeconds: 15
 
+stoClientConfig:
```
Comment (kapil-harness) [inline 120-ng-manager/config.yml:145]: lets add its corresponding replace_configs too
`dockerization/ng-manager/scripts/replace_configs.sh`

Also, add these configs in all the envs in ng-prod-manifest

**Review by kapil-harness — CHANGES_REQUESTED** [review]

**Review by kapil-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52046)* status: 4 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java` line 104 at r1](https://reviewable.io/reviews/harness/harness-core/52046#-NcfuRbl87ARpqJfX0xQ:-NcfuRbl87ARpqJfX0xR:bwkcsiz) ([raw file](https://github.com/harness/harness-core/blob/6025f3e3e6de2150c08cff46fd667858950ef528/120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java#L104)):*
> ```Java
> 
>   private String getBaseUrl(String moduleName) throws IOException {
>     String finalBaseUrl = "";
> ```

we don't need to create string here and then return later, each switch case can independently return



_Code snippet:_
```Java
software.wings.beans.loginSettings.outbox.LoginSettingsOutboxEventHandler#handle
```
___
*[`120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java` line 105 at r1](https://reviewable.io/reviews/harness/harness-core/52046#-NcfujU3D82m00Mu03qe:-NcfujU3D82m00Mu03qf:b-k2m2hv) ([raw file](https://github.com/harness/harness-core/blob/6025f3e3e6de2150c08cff46fd667858950ef528/120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java#L105)):*
> ```Java
>   private String getBaseUrl(String moduleName) throws IOException {
>     String finalBaseUrl = "";
>     try {
> ```

try catch is not required here
___
*[`120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java` line 106 at r1](https://reviewable.io/reviews/harness/harness-core/52046#-Ncfuek04O5fBp09HoHI:-Ncfuek04O5fBp09HoHJ:bdwpolg) ([raw file](https://github.com/harness/harness-core/blob/6025f3e3e6de2150c08cff46fd667858950ef528/120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java#L106)):*
> ```Java
>     String finalBaseUrl = "";
>     try {
>       if (PLATFORM.equals(moduleName)) {
> ```

why is this not part of the switch case?
___
*[`120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java` line 214 at r1](https://reviewable.io/reviews/harness/harness-core/52046#-NcfupzsDPUmZZ9eQwUm:-Ncfupzt0HzU0aJktoXF:bnj7hl0) ([raw file](https://github.com/harness/harness-core/blob/6025f3e3e6de2150c08cff46fd667858950ef528/120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java#L214)):*
> ```Java
>     JSONObject jsonObject = new JSONObject(responseString);
>     
>     String finalVersion = "";
> ```

same as above

File: 120-ng-manager/config.yml (hunk 1 above)
Comment (r0610205) [reply 120-ng-manager/config.yml:145]: And to corresponding helm chart

File: 120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java [hunk 2]
```diff
@@ -96,25 +100,49 @@ private void checkVersionChange() {
...
+        case CF:
+          finalBaseUrl = new StringBuilder(nextGenConfiguration.getFfServerClientConfig().getBaseUrl()).toString();
+          break;
+        case CI:
+          finalBaseUrl = nextGenConfiguration.getCiManagerClientConfig().getBaseUrl();
+          break;
+        case SRM:
+          finalBaseUrl = nextGenConfiguration.getCvngClientConfig().getBaseUrl();
+          break;
+        case STO:
```
Comment (r0610205) [inline 120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java:132]: We have two components in STO, STO Manager & STO Core, we only pulling one version?

**Review by neelam-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52046)* status: Some of the checks are too old. 4 file(s) to review. 1 change requests. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/config.yml` line 145 at r1](https://reviewable.io/reviews/harness/harness-core/52046#gh-1305263274:-Ncyj6g246hsKNtyHPCj:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/6025f3e3e6de2150c08cff46fd667858950ef528/120-ng-manager/config.yml#L145)):*
<details><summary><i>Previously, r0610205 (Sergey Bobrov) wrote…</i></summary><blockquote>

And to corresponding helm chart 
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java` line 104 at r1](https://reviewable.io/reviews/harness/harness-core/52046#-NcfuRbl87ARpqJfX0xQ:-NciBrZ33fPUOc9vfbfq:b-oqdfl8) ([raw file](https://github.com/harness/harness-core/blob/6025f3e3e6de2150c08cff46fd667858950ef528/120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java#L104)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

we don't need to create string here and then return later, each switch case can independently return
</blockquote></details>

Both approaches work fine imho. The rationale for creating string here was to return `empty` in the end if none of the cases matched and populated an appropriate baseUrl. Can just do an early return. Thanks
___
*[`120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java` line 105 at r1](https://reviewable.io/reviews/harness/harness-core/52046#-NcfujU3D82m00Mu03qe:-NciHelL7iXWk8YpzNvF:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/6025f3e3e6de2150c08cff46fd667858950ef528/120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java#L105)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

try catch is not required here
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java` line 106 at r1](https://reviewable.io/reviews/harness/harness-core/52046#-Ncfuek04O5fBp09HoHI:-NciCYkqDhOrY1gjrKpp:b-21rubv) ([raw file](https://github.com/harness/harness-core/blob/6025f3e3e6de2150c08cff46fd667858950ef528/120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java#L106)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

why is this not part of the switch case?
</blockquote></details>

because it's a special case and is not an independent moduleType so keeping it separate . https://github.com/harness/harness-core/blob/c417ccf6457af91a1dd38e0b646a21b3ef2eb2db/970-ng-commons/src/main/java/io/harness/entities/ModuleType.java#L26 
___
*[`120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java` line 132 at r1](https://reviewable.io/reviews/harness/harness-core/52046#gh-1305937095:-NcyjRhR5R52gskBpBZp:bj4q61s) ([raw file](https://github.com/harness/harness-core/blob/6025f3e3e6de2150c08cff46fd667858950ef528/120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java#L132)):*
<details><summary><i>Previously, r0610205 (Sergey Bobrov) wrote…</i></summary><blockquote>

We have two components in STO, STO Manager & STO Core, we only pulling one version?
</blockquote></details>

Yes that's right. As per guideline from product we are only pulling version for STO Core. 
___
*[`120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java` line 214 at r1](https://reviewable.io/reviews/harness/harness-core/52046#-NcfupzsDPUmZZ9eQwUm:-NcijqfN7y3PFE4cEzOh:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/6025f3e3e6de2150c08cff46fd667858950ef528/120-ng-manager/src/main/java/io/harness/ng/moduleversioninfo/runnable/UpdateVersionInfoTask.java#L214)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.

File: 120-ng-manager/config.yml (hunk 1 above)
Comment (neelam-harness) [reply 120-ng-manager/config.yml:145]: Added. Creating a separate PR for the ng-prod-manifest change. Thanks for pointing this out!

**Pull request closed without merging**